src/modules/roc_fec/writer.cpp

```diff
@@ -68,51 +69,82 @@ void Writer::write(const packet::PacketPtr& pp) {
         } while (source_ == pp->rtp()->source);
     }
 
+    encode_source_packet_(*pp->fec());
+    write_source_packet_(pp);
+
+    cur_packet_++;
+
+    if (cur_packet_ == cur_sblen_) {
+        handle_next_block_();
+    }
+}
+
+void Writer::resize(size_t sblen) {
+    roc_log(LogDebug, "fec writer: update sblen, cur=%lu next=%lu",
+            (unsigned long)cur_sblen_, (unsigned long)sblen);
+
+    next_sblen_ = sblen;
+}
+
+void Writer::encode_source_packet_(const packet::FEC& fec) {
+    if (cur_packet_ == 0) {
+        if (!encoder_.begin(cur_sblen_, cur_rblen_)) {
```

Comment:
I do not like this approach. `begin_block_` is an extra wrapper around `encoder.begin()` that only adds an error message to the log and no more.

I understand that you want to see actions that we perform when the block begins and actions when the blog ends.

Why you do not like the separation of encoding and writing?

Reply:
> Why you do not like the separation of encoding and writing?

I'm OK with that, we can keep them in separate methods. E.g., you can extract the fill-and-compose part into a new method and we'll get something like:

```
    encoder_.set(cur_packet_, fec.payload);
    fill_packet_(pp);
    writer_.write(pp);
```

> begin_block_ is an extra wrapper around encoder.begin() that only adds an error message to the log and no more.

Yes, but I think it's not that bad because:

* Writer::begin_block_ and encoder.begin() are different abstraction layers. It's usually not recommended to mix different abstraction layers (google for "single level of abstraction principle") because it makes the code harder to understand.

* This wrapper removes some boilerplate (logging) from the Writer::write which allows us to keep that method more concise and high-level and the begin/fill/end flow becomes more clear.

If you still don't think this would be more readable, let's leave this as is. But in this case I suggest at least to hide the low-level code that increments and check the packet number:

```
    cur_packet_++;

    if (cur_packet_ == cur_sblen_) {
        handle_next_block_();
    }
```

inside the handle_next_block_ method (and maybe rename it accordingly).

Reply:
I have reworked it a little bit. I still do not like it but it seems better than it was...

Reply:
Looks pretty good.